packages/blob/README.md

```diff
@@ -202,6 +220,124 @@ export async function POST(request: Request) {
 }
 ```
 
+### Uploading directly from browsers
+
+The above example uploads a file through a vercel route. This solution is limited to a 4Mb file size. In order to bypass this limit, it is possible to upload a file directly from within the client, after generating a single-use token.
+
+```tsx
+// /app/UploadForm.tsx
+
+'use client';
+
+import type { BlobResult, put } from '@vercel/blob';
+import { useState } from 'react';
+
+export default function UploadForm() {
+  const inputFileRef = useRef<HTMLInputElement>(null);
+  const [blob, setBlob] = useState<BlobResult | null>(null);
+  return (
+    <>
+      <h1>App Router Client Upload</h1>
+
+      <form
+        onSubmit={async (event): Promise<void> => {
+          event.preventDefault();
+
+          const file = inputFileRef.current?.files?.[0];
+          if (!file) {
+            return;
+          }
+
+          const clientTokenData = (await fetch('/api/upload-token', {
+            method: 'POST',
+            body: JSON.stringify({
+              pathname: file.name,
+            }),
+          }).then((r) => r.json())) as { clientToken: string };
+
+          const blobResult = await put(file.name, file, {
+            access: 'public',
+            token: clientTokenData.clientToken,
+          });
+
+          setBlob(blobResult);
+        }}
+      >
+        <input name="file" ref={inputFileRef} type="file" />
+        <button type="submit">Upload</button>
+      </form>
+      {blob && (
+        <div>
+          Blob url: <a href={blob.url}>{blob.url}</a>
+        </div>
+      )}
+    </>
+  );
+}
+```
+
+```ts
+// /app/api/upload-token/route.ts
+
+import {
+  generateClientTokenFromReadWriteToken,
+  type GenerateClientTokenOptions,
+} from '@vercel/blob';
+import { NextResponse } from 'next/server';
+
+export async function POST(request: Request): Promise<NextResponse> {
+  // On a real website, this route would be protected by authentication, see: https://nextjs.org/docs/pages/building-your-application/routing/authenticating
+  // Here, we accept the `pathname` from the browser, but in some situations, you may even craft the pathname
+  // based on the authentication result
+  const body = (await request.json()) as { pathname: string };
+
+  return NextResponse.json({
+    clientToken: await generateClientTokenFromReadWriteToken({
+      ...body,
+      onUploadCompleted: {
+        callbackUrl: `https://${
+          process.env.VERCEL_URL ?? ''
+        }/api/upload-completed`,
+        metadata: JSON.stringify({ foo: 'bar' }),
+      },
+      maximumSizeInBytes: 10_000_000, // 10 Mb
+      allowedContentTypes: 'text/plain',
+    }),
+  });
+}
+```
+
+```ts
+// /app/api/upload-completed/route.ts
```

Comment:
Should we be explicit and call this route `/app/api/file-upload-completed/route.ts`?